File: modules/core/native/macos/iosurface.cpp
//
// Created by Nikita Shtengauer on 20.11.2023.
//

#include "openglfx-macos.h"

#include <IOSurface/IOSurface.h>
#include <OpenGL/OpenGL.h>




jni_win_iosurface(jlong, nCreateIOSurface)(JNIEnv *, jobject, jint width, jint height) {
    int bytes = 4;
    CFStringRef keys[3]{
            kIOSurfaceWidth, kIOSurfaceHeight, kIOSurfaceBytesPerElement
    };
    CFNumberRef values[3]{
            CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &width),
            CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &height),
            CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &bytes)
    };

    CFDictionaryRef dictionary = CFDictionaryCreate(
            nullptr,
            (const void **) keys, (const void **) values, 3,
            &kCFCopyStringDictionaryKeyCallBacks,
            &kCFTypeDictionaryValueCallBacks);
    IOSurfaceRef ioSurfaceRef = IOSurfaceCreate(dictionary);
    CFRelease(dictionary);
    return (jlong) ioSurfaceRef;
}

jni_win_iosurface(void, nReleaseIOSurface)(JNIEnv *, jobject, jlong ioSurfaceRef) {
    CFRelease((IOSurfaceRef) ioSurfaceRef);
}

jni_win_iosurface(jint, nCGLTexImageIOSurface2D)(JNIEnv *, jobject,
                                      jlong ctx, jint target, jint internalFormat, jint width, jint height, jint format,
                                      jint type,
                                      jlong ioSurfaceRef, jint plane) {
    return (jint) CGLTexImageIOSurface2D(
            (CGLContextObj) ctx,
            (GLenum) target, (GLenum) internalFormat,
            (GLsizei) width, (GLsizei) height,
            (GLenum) format, (GLenum) type,
            (IOSurfaceRef) ioSurfaceRef, (GLuint) plane);
}

jni_win_iosurface(void, nIOSurfaceLock)(JNIEnv *, jobject, jlong ioSurfaceRef) {
    IOSurfaceLock((IOSurfaceRef) ioSurfaceRef, kIOSurfaceLockReadOnly, nullptr);
}

jni_win_iosurface(void, nIOSurfaceUnlock)(JNIEnv *, jobject, jlong ioSurfaceRef) {
    IOSurfaceUnlock((IOSurfaceRef) ioSurfaceRef, kIOSurfaceLockReadOnly, nullptr);
}

File: modules/core/kotlin/com/huskerdev/openglfx/internal/platforms/macos/IOSurface.kt
package com.huskerdev.openglfx.internal.platforms.macos

import com.huskerdev.grapl.gl.GLContext
import com.huskerdev.openglfx.GLExecutor.Companion.glBindTexture
import com.huskerdev.openglfx.GLExecutor.Companion.glGenTextures
import com.huskerdev.openglfx.GL_BGRA
import com.huskerdev.openglfx.GL_RGBA
import com.huskerdev.openglfx.GL_TEXTURE_RECTANGLE
import com.huskerdev.openglfx.GL_UNSIGNED_INT_8_8_8_8_REV

internal class IOSurface(val width: Int, val height: Int) {

    companion object {
        @JvmStatic private external fun nCreateIOSurface(width: Int, height: Int): Long
        @JvmStatic private external fun nReleaseIOSurface(ioSurfaceRef: Long)
        @JvmStatic private external fun nCGLTexImageIOSurface2D(
            ctx: Long, target: Int, internalFormat: Int,
            width: Int, height: Int, format: Int, type: Int,
            ioSurfaceRef: Long, plane: Int): Int
        @JvmStatic private external fun nIOSurfaceLock(ioSurfaceRef: Long)
        @JvmStatic private external fun nIOSurfaceUnlock(ioSurfaceRef: Long)
    }

    private val handle = nCreateIOSurface(width, height)

    fun createBoundTexture(): Int {
        val texture = glGenTextures()
        glBindTexture(GL_TEXTURE_RECTANGLE, texture)
        nCGLTexImageIOSurface2D(
            GLContext.current().handle,
            GL_TEXTURE_RECTANGLE,
            GL_RGBA,
            width, height,
            GL_BGRA,
            GL_UNSIGNED_INT_8_8_8_8_REV,
            handle, 0)
        return texture
    }

    fun lock() = nIOSurfaceLock(handle)

    fun unlock() = nIOSurfaceUnlock(handle)

    fun dispose() = nReleaseIOSurface(handle)
}